compiler/src/lnntoamm/Types.ts

```diff
@@ -85,54 +103,85 @@ export class Builtin extends Type {
     return this;
   }
 
-  compatibleWithConstraint(ty: Type): boolean {
+  compatibleWithConstraint(ty: Type, scope: Scope): boolean {
     if (ty instanceof Builtin) {
       return this.eq(ty);
-    } if (ty instanceof OneOf || ty instanceof Generated) {
-      return ty.compatibleWithConstraint(this);
+    } else if (ty instanceof OneOf || ty instanceof Generated) {
+      return ty.compatibleWithConstraint(this, scope);
+    } else if (ty instanceof HasOperator) {
+      return Has.operator(ty, scope, this).length !== 0;
+    } else if (ty instanceof HasMethod) {
+      return Has.method(ty, scope, this).length !== 0;
+    } else if (ty instanceof HasField) {
+      return false;
+    } else if (ty instanceof Interface) {
+      return ty.fields.length === 0 &&
+            ty.methods.every(m => this.compatibleWithConstraint(m, scope)) &&
+            ty.operators.every(o => this.compatibleWithConstraint(o, scope));
     } else {
       TODO('other types constraining to builtin types');
     }
   }
 
-  constrain(ty: Type) {
-    if (ty instanceof OneOf || ty instanceof Generated) {
-      ty.constrain(this);
-    } else if (!this.compatibleWithConstraint(ty)) {
+  constrain(ty: Type, scope: Scope) {
+    if (ty instanceof OneOf || ty instanceof Generated || ty instanceof Interface) {
+      ty.constrain(this, scope);
+    } else if (ty instanceof HasOperator) {
+      if (Has.operator(ty, scope, this).length === 0) {
+        throw new Error(`type ${this.name} does not have operator ${ty.name}`);
+      }
+    } else if (ty instanceof HasMethod) {
+      if (Has.method(ty, scope, this).length === 0) {
+        throw new Error(`type ${this.name} does not have method ${ty.name}(${ty.params.map(p => p === null ? this : p).map(ty => ty.name).join(', ')})`);
+      }
+    } else if (ty instanceof HasField) {
+      throw new Error(`type ${this.name} does not have field ${ty.name}`);
+    } else if (!this.compatibleWithConstraint(ty, scope)) {
       throw new Error(`type ${this.name} could not be constrained to ${ty.name}`);
     }
   }
 
   eq(that: Equalable): boolean {
-    return that instanceof Builtin && this.ammName === that.ammName;
+    if (that instanceof Builtin) {
+      return this.ammName === that.ammName;
+    } else if (that instanceof Interface) {
+      if (that instanceof Generated && that.delegate !== null) {
+        return this.eq(that.delegate);
+      }
+      return that.tempDelegate !== null && this.eq(that.tempDelegate);
```

Comment:
What exactly is the difference between `.delegate` and `.tempDelegate`?

Reply:
delegates are only set once, while `tempDelegates` can be reset. I think only `Interface` is necessary (and we can get rid of `Generated`) but I want to wait until untyped fn parameters are implemented before making that decision (since i'm not 100% certain that they can be represented with just an `Interface`)

Reply:
What do you think would be needed extra to represent them versus `Interface`s? It might make sense to just add that as a potentially explicit syntax to `Interface` as well to force them to be the same (and also make them both equally expressive, I would hate if certain code can only be written in explicit or implicit style only and not the other).

Reply:
Well right now there would be issues with type inference - *at least* fn selection will break, so the PR for fixing that would be the earliest that I can consider getting rid of the `Generated` type. It does make sense, but I'll most likely have to generate constraints such as `variableType.constrain(Type.hasMethod(...))`, etc